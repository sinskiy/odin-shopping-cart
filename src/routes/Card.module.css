.card {
  width: fit-content;
  background-color: var(--surface);
  border-radius: var(--rounded-card);
  padding-bottom: 2rem;
}

.card-image {
  border-radius: var(--rounded-card);
}

.card-content {
  margin-top: 1rem;
  margin-left: 1rem;
  margin-bottom: 2rem;
}
.card-price {
  font-size: 2rem;
  font-weight: 600;
}
.card-title {
  margin-top: 0.5rem;
  font-size: 1rem;
  font-weight: 400;
}

.card-button {
  margin-left: 1rem;
  margin-right: 1rem;
  margin-top: 1rem;
}
.card-button button {
  width: 100%;
}
