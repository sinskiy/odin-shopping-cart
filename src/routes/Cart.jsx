import { useLoaderData } from "react-router-dom";
import { getCart } from "../cart";
import { getCards } from "../cards";
import CartEntry from "../components/CartEntry";

export async function cartLoader() {
  const cards = await getCards();

  const cart = getCart();
  const cartEntries = Object.entries(cart);
  return { cards, cartEntries };
}

export default function Cart() {
  const { cards, cartEntries } = useLoaderData();
  const cartCards = cartEntries?.map(([id, amount]) => {
    return { ...cards.find((card) => card.id == id), amount: amount };
  });
  return (
    <section>
      {cartCards.map((card) => (
        <article key={card.id}>
          <CartEntry imageUrl={card.image} {...card} />
        </article>
      ))}
    </section>
  );
}
