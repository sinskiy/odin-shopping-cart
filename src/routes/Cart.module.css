.total {
  display: flex;
  justify-content: space-between;
  align-items: center;

  background-color: var(--surface);
  padding: 2rem;
  border-radius: var(--rounded-card);

  margin-bottom: 2rem;
}
.total h2 {
  font-weight: 400;
}
