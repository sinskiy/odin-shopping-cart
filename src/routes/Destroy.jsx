import { redirect } from "react-router-dom";
import { deleteCartEntry } from "../cart";

export function destroyAction({ params }) {
  deleteCartEntry(params.cardId);
  return redirect("/cart");
}
