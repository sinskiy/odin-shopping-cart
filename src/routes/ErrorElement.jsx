import { useNavigate, useRouteError } from "react-router-dom";
import { errorElement, nav } from "./ErrorElement.module.css";

export default function ErrorElement() {
  const navigate = useNavigate();

  const error = useRouteError();
  console.error(error);

  return (
    <section className={errorElement}>
      <p>unexpected error</p>
      <nav className={nav}>
        <button onClick={() => navigate(-1)} className="styled secondary">
          go back
        </button>
        <button onClick={() => navigate("/")} className="styled primary">
          go home
        </button>
      </nav>
    </section>
  );
}
