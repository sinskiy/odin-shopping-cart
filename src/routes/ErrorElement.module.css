.error-element {
  width: fit-content;
  margin: auto;
  text-align: center;
}

.nav {
  margin-top: 1rem;

  display: flex;
  gap: 0.5rem;
}
