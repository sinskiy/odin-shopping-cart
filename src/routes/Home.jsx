import { hero, heroText } from "./Home.module.css";

export default function Home() {
  return (
    <>
      <section className={hero}>
        <img
          src="https://images.ctfassets.net/u41cm62nxtp7/2oQgUuzHo1oV3nazkQVaOI/85ca99ce0ad2a1db6117dd889554a7ce/img-hero-teaser-history_2x.jpg?fm=webp"
          alt=""
        />
        <div className={heroText}>
          <h1>
            high quality.
            <br />
            low prices.
          </h1>
        </div>
      </section>
      <section>
        <h2>this is not a real website</h2>
        <div className="text">
          <p>
            Welcome to Lidl US. Here, we believe high quality shouldn’t have to
            come at a premium price. Throughout our stores, you’ll find local
            produce, fresh meats and seafood, imported cheeses, European
            pastries, weekly surprises and more at unbelievably low prices.
            Sound too good to be true? Maybe. But at Lidl, you really can have
            your cake and eat it too. Here’s why:
          </p>
          <p>
            Our selection is expertly curated. Instead of offering countless
            brands in every category, we work closely with vendors to source
            high-quality products at competitive prices.
          </p>
          <p>
            You’re busy. We get it. Every store features an intuitive layout for
            a seamless experience.
          </p>
          <p>
            Our private label products are a result of meticulous sourcing. We
            partner with local and regional farms and suppliers to give you
            top-tier freshness and quality. However, we wouldn’t be Lidl without
            a delicious international selection. From authentic Greek pastries
            to real Spanish olives, imported specialties are always waiting in
            stores.
          </p>
          <p>
            Around 80% of our products are private label—all of which undergo
            rigorous taste, quality and sensory testing. What can we say, we’re
            picky.
          </p>
          <p>
            We’re more than just food. Every Wednesday, new surprises await in
            the middle of our stores. You’ll find anything from comfy sweatpants
            to cast iron cookware and more.
          </p>
          <p>
            We keep prices low by keeping overhead to a minimum and eliminating
            any unnecessary costs. Take our “bring your own bag” policy. Not
            only is the cost of bags not built into our product prices, but this
            approach is a greener solution—something we value as a company.
            Also, instead of restocking shelves item by item, products are
            stored in the boxes they arrive in to save time.
          </p>
        </div>
      </section>
    </>
  );
}
