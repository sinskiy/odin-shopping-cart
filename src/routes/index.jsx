import Cart from "./Cart";
import Home from "./Home";
import Root from "./Root";
import Shop from "./Shop";

const routes = [
  {
    path: "/",
    element: <Root />,
    children: [
      { index: true, element: <Home /> },
      {
        path: "/shop",
        element: <Shop />,
      },
      {
        path: "/cart",
        element: <Cart />,
      },
    ],
  },
];

export default routes;
