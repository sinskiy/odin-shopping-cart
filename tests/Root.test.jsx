import { render, screen } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { describe, expect, it } from "vitest";
import routes from "../src/routes";
import userEvent from "@testing-library/user-event";

describe("Root component", () => {
  it("is the same as a snapshot", () => {
    const router = createMemoryRouter(routes);
    const { container } = render(<RouterProvider router={router} />);

    expect(container).toMatchSnapshot();
  });
  it("changes routes correctly", async () => {
    const user = userEvent.setup();
    const router = createMemoryRouter(routes);
    render(<RouterProvider router={router} />);

    const shopNavLink = screen.getByRole("link", { name: /shop$/i });

    await user.click(shopNavLink);

    screen.debug();
    expect(shopNavLink).toHaveClass("active");
  });
});
