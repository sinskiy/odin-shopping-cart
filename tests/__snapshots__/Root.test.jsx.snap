// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Root component > is the same as a snapshot 1`] = `
<div>
  <header>
    <a
      href="/"
    >
      Shopping cart
    </a>
    <nav>
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        Home
      </a>
      <a
        class=""
        href="/shop"
      >
        Shop
      </a>
    </nav>
  </header>
  <main>
    <div>
      home
    </div>
  </main>
</div>
`;
